File: app/src/main/java/io/github/zwieback/familyfinance/app/AbstractApplication.java
package io.github.zwieback.familyfinance.app;

import android.app.Application;
import android.support.annotation.NonNull;

import java.sql.Connection;

import io.github.zwieback.familyfinance.app.lifecycle.creator.DatabaseViewCreator;
import io.github.zwieback.familyfinance.core.database.exception.SQLException;
import io.requery.Persistable;
import io.requery.android.sqlite.DatabaseProvider;
import io.requery.reactivex.ReactiveEntityStore;
import io.requery.reactivex.ReactiveSupport;
import io.requery.sql.Configuration;
import io.requery.sql.EntityDataStore;

public abstract class AbstractApplication extends Application {

    protected static final int DB_VERSION = 1;

    private ReactiveEntityStore<Persistable> dataStore;

    /**
     * @return {@link EntityDataStore} single instance for the application.
     * <p/>
     * Note if you're using Dagger you can make this part of your application level module returning
     * {@code @Provides @Singleton}.
     */
    public final ReactiveEntityStore<Persistable> getData() {
        if (dataStore == null) {
            DatabaseProvider databaseProvider = buildDatabaseProvider();
            Configuration configuration = databaseProvider.getConfiguration();
            dataStore = ReactiveSupport.toReactiveStore(
                    new EntityDataStore<Persistable>(configuration));
        }
        return dataStore;
    }

    @NonNull
    protected abstract DatabaseProvider buildDatabaseProvider();

    /**
     * Workaround for creating views, while requery does not support creating views.
     * <p>
     * Note: This method must be called before the creation of the {@link #dataStore},
     * otherwise {@link java.sql.SQLException} will be thrown.
     *
     * @see <a href="https://github.com/requery/requery/issues/721#issuecomment-344153774">
     * Using SQLite views
     * </a>
     */
    protected final void createViews(Configuration configuration) {
        try (Connection connection = configuration.getConnectionProvider().getConnection()) {
            new DatabaseViewCreator(connection).createViews();
        } catch (java.sql.SQLException e) {
            throw new SQLException(e.getMessage(), e);
        }
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/app/AbstractApplication.kt
package io.github.zwieback.familyfinance.app

import androidx.multidex.MultiDexApplication
import com.jakewharton.threetenabp.AndroidThreeTen
import com.mikepenz.iconics.Iconics
import io.github.zwieback.familyfinance.app.lifecycle.creator.DatabaseViewCreator
import io.github.zwieback.familyfinance.app.lifecycle.destroyer.DatabaseViewDestroyer
import io.github.zwieback.familyfinance.core.model.Models
import io.requery.Persistable
import io.requery.android.sqlite.DatabaseProvider
import io.requery.android.sqlitex.SqlitexDatabaseSource
import io.requery.reactivex.ReactiveEntityStore
import io.requery.reactivex.ReactiveSupport
import io.requery.sql.Configuration
import io.requery.sql.EntityDataStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import java.sql.SQLException

abstract class AbstractApplication : MultiDexApplication() {

    /**
     * Note if you're using Dagger you can make this part of your application
     * level module returning `@Provides @Singleton`.
     *
     * @return [EntityDataStore] single instance for the application.
     */
    val data: ReactiveEntityStore<Persistable> by lazy {
        val databaseProvider = createDatabaseProvider()
        setupDatabaseProvider(databaseProvider)
        val configuration = databaseProvider.configuration
        destroyViews(configuration)
        createViews(configuration)
        ReactiveSupport.toReactiveStore(EntityDataStore(configuration))
    }

    override fun onCreate() {
        super.onCreate()
        AndroidThreeTen.init(this)
        Iconics.init(this)
    }

    private fun createDatabaseProvider(): DatabaseProvider<*> {
        return SqlitexDatabaseSource(this, Models.DEFAULT, DB_VERSION)
    }

    protected open fun setupDatabaseProvider(source: DatabaseProvider<*>) {
        // do nothing
    }

    /**
     * Workaround for creating views, while requery does not support
     * creating views.
     *
     * Note: This method must be called before the creation of the [data],
     * otherwise [SQLException] will be thrown.
     *
     * See [Using SQLite views](https://github.com/requery/requery/issues/721.issuecomment-344153774)
     */
    @Throws(SQLException::class)
    private fun createViews(configuration: Configuration) = runBlocking(Dispatchers.IO) {
        configuration.connectionProvider.connection.use { connection ->
            DatabaseViewCreator(connection).createViews()
        }
    }

    /**
     * Workaround for destroying views, while requery does not support
     * destroying views.
     *
     * Note: This method must be called before the creation of the [data],
     * otherwise [SQLException] will be thrown.
     *
     * See [Using SQLite views](https://github.com/requery/requery/issues/721.issuecomment-344153774)
     */
    @Throws(SQLException::class)
    private fun destroyViews(configuration: Configuration) = runBlocking(Dispatchers.IO) {
        configuration.connectionProvider.connection.use { connection ->
            DatabaseViewDestroyer(connection).destroyViews()
        }
    }

    companion object {
        const val DB_VERSION = 9
    }
}
